### PR TITLE
Unused refinements NULL, no error on NULL via WORD!/PATH!

### PR DESCRIPTION
**Pull request by hostilefork**
[This has been pondered for many months, long-winded discussion here.](https://forum.rebol.info/t/are-nulls-the-best-representation-for-unused-refinements/1140)  This commit is not a done deal--there's pros and cons--but there's certainly things to like about the implementation.  I'd like to review it in a broader context and understand if it's really what we want to finalize on, or if there is a better answer in here somewhere.

The PR title says the most important bit.  Where unused refinements used to be BLANK!, this makes them NULL...and then gets rid of the idea that the evaluator requires a GET-WORD! or GET-PATH! to fetch nulls without error.  This tightens things up, for instance in places like ARRAY/INITIAL where you might literally want to ask for BLANK!, and where not saying `/INITIAL` might have a different subtlety to it.

Most code should work the same, though anything which detected a type and expected BLANK! will get in trouble.  OPT on a NULL gives you a VOID!, so there may be some superfluous OPT to take out.  You'll probably find you have a lot of GET-WORD!s in places they're not really needed now, and feel the code is cleaner by dropping the colons.

To keep this from getting too permissive, it means that words that are bound to things that have never been assigned (e.g. not function arguments) must be set to VOID! to trigger errors when they are accessed.  So the system does this in a way that's very similar to historical UNSET!...but I still prefer to say that NULL variables are "unset" because *they have no value*.  This PR tries to inject the terminology that VOID! variables are instead **undefined?** (but "voided?" might be better, I just thought we'd try it).

I like VOID because it's the return value of functions that aren't supposed to have any usable result.  But more accurately, think of a paper check that has VOID written on it.  You can still hand it to people, and you can get in trouble if you try to cash it.  But you can't cause a problem by trying to cash a NULL check because there's no paper and nothing written on it--it is the literal absence of a check.

Note that also in this commit is the ability to use NULL in evaluated refinements.  You can write **append/(if false [/only]) [a b c] [d e]** and it works.  Please note that it might be `@only` and not `/only` in the future, but I'm reasonably confident that some chaining concept will be used...and the [unused no-argument refinement state will stay NULL as well](https://forum.rebol.info/t/should-refinement-arguments-be-refinement-if-used/735/3?u=hostilefork).

### CODE REVIEW

**Comment by hostilefork** [comment]
Ok...having had a couple of months to reflect on this, I think I've come to believe it is the right answer (!)

I had hoped to have more codebases in Travis and running tests before making such changes.  That way I could fix the affected code that others have.  **But I cannot wait forever on that.  :-/  If you have trouble with this, please let me know and I will help.**

The easiest part of rolling with this change involves taking cases where refinements were explicitly tested for BLANK! and switching that to a test for **null? ref** or **null = type of ref** (or **set? 'ref**).  This hopefully won't be too onerous.

Maybe trickier to deal with is that the notion of "unset" no longer conflates variables that contain no value (nulled) with those that are VOID!.  This is a good thing, because having null refinements is a continuation of the trend that null is commonplace and likely to be intentional.  Hence we don't want typos like **:some-varable** to be fetched as a now-"friendly" null.  It's better as a less-friendly VOID! (or an even less friendly unbound, which was the trend in MODULE!s for typos, at the cost of forcing you to name your declarations up front).

Anyway: this cements the idea that there is not any distinction between a NULL refinement and no refinement used at all.  The clearest place where this causes hiccups is in specialization, when you need a way to remove a refinement from the interface -but- also not use it.  One idea for notationally how this might be tackled is something like:

    append-no-only: specialize 'append/only [only: null]

In other words, when you use a refinement in the specialization like that, it still gets bound in the body block...where you can choose to override its value with NULL (or take the default assumption that you meant "use it").  This could also apply to non-refinements you want to set to NULL as well.

Please let me know about any concerns with this PR, as I intend to push it soon.

**Comment by lkppo** [comment]
I would see go ahead.

Didi you receive my email to brian@hostilefork.com about the OpenBSD Server ?
